core/consensus/synchronizer/impl/synchronizer_impl.cpp

```diff
@@ -0,0 +1,214 @@
+                         network::Direction::DESCENDING,
+                         boost::none};
+    requestBlocks(std::move(request), peer, std::move(cb));
+  }
+
+  void SynchronizerImpl::requestBlocks(const libp2p::peer::PeerInfo &peer,
+                                       const primitives::BlockHash &hash,
+                                       RequestCallback cb) {
+    auto request_id = last_request_id_++;
+    // using last_finalized, because if the block, which we want to get, is in
+    // non-finalized fork, we are not interested in it; otherwise, it 100% will
+    // be a descendant of the last_finalized
+    BlockRequest request{request_id,
+                         BlockRequest::kBasicAttributes,
+                         block_tree_->getLastFinalized(),
+                         hash,
+                         network::Direction::DESCENDING,
+                         boost::none};
+    requestBlocks(std::move(request), peer, std::move(cb));
+  }
+
+  void SynchronizerImpl::requestBlocks(network::BlockRequest request,
+                                       const libp2p::peer::PeerInfo &peer,
+                                       RequestCallback cb) const {
+    auto peer_client_it = network_state_->peer_clients.find(peer.id);
+    if (peer_client_it == network_state_->peer_clients.end()) {
+      log_->info("no peer with id {}", peer.id.toBase58());
+      return cb(Error::NO_SUCH_PEER);
+    }
+
+    peer_client_it->second->blocksRequest(
+        std::move(request),
+        [self{shared_from_this()},
+         cb = std::move(cb)](auto &&block_response_res) mutable {
+          self->handleBlocksResponse(
+              std::forward<decltype(block_response_res)>(block_response_res),
+              cb);
+        });
+  }
+
+  void SynchronizerImpl::handleBlocksResponse(
+      const outcome::result<network::BlockResponse> &response_res,
+      const RequestCallback &cb) const {
+    if (!response_res) {
+      return cb(response_res.error());
+    }
+
+    for (auto &block_data : response_res.value().blocks) {
+      auto block_opt = block_data.toBlock();
+      if (!block_opt) {
+        // there is no sense in continuing: blocks are coming in
+        // sequential order, and if one of them is skipped, we will
+        // not be able to insert any further
+        return cb(Error::INVALID_BLOCK);
+      }
+      auto block_insert_res = block_tree_->addBlock(std::move(*block_opt));
+      if (!block_insert_res
+          && block_insert_res.error()
+                 != blockchain::BlockTreeError::BLOCK_EXISTS) {
+        // if such block already exists, it's OK
+        return cb(block_insert_res.error());
+      }
+    }
+
+    cb(outcome::success());
+  }
+
+  outcome::result<network::BlockResponse> SynchronizerImpl::processRequest(
+      const BlockRequest &request) const {
+    BlockResponse response{request.id};
+
+    // firstly, check if we have both "from" & "to" blocks (if set)
```

Comment:
`firstly`, `secondly` and `thirdly` parts can be put into separate functions for better readability